beam-sync.md

```diff
@@ -0,0 +1,187 @@
+
+#### Optimization 3: Speculative Execution
+
+Many transactions included in the same block access independent parts of the state. So we can speculatively execute the transactions out-of-order. To speculatively execute, we treat each transaction as if it was the first & only transaction in the block. The transaction may fail, or execute on a different code path, in which case we get no benefit. But for the many state-independent transactions, we can correctly collect the required JIT-data for each transaction in parallel.
+
+Note: if the same block includes multiple transactions from the same sender, then we group them together to run in serial. If you don't group them together, then the 2nd to nth transaction from the sender would always fail the nonce check, and provide no benefit.
+
+There are probably other games you could play to predict transactions that depend on each other, and serialize them. But it hasn't been explored yet.
+
+## Network Health
+
+There are a few benefits to the health of the network.
+
+One was already mentioned: Beam Syncing peers can serve data to other Beam Syncing peers. This means that a giant swarm of newly-connected nodes do not choke each other from joining the network.
+
+Additionally, there is the cache-friendliness of the design. Storage I/O dominates the cost of serving trie nodes. Since peers are serving the same data that they just received, it should typically be in cache, whether that's at the application level, DB level, or OS level. In fact, since the trie nodes were already accessed for blocks that were recently executed, nodes that are unaware of Beam Sync might even have these trie nodes cached. This reduces the strain of serving state data.
+
+(Fast Sync, by comparison, seems to be purpose-built to wreck the cache, by forcing a node to do a complete read out of all trie data from the database)
+
+## Backfill coordination
+
+Which brings us to the backfill strategy. The most naive implementation is to walk the trie, requesting all missing nodes. This is basically Fast Sync. Naively walking the trie, where each client picks its own scanning strategy, has a couple drawbacks:
+
+- Beam-syncing peers can't assist each other in backfilling data
+- It breaks the cache-friendliness of JIT-data collection
+
+One approach to address these issues is to coordinate clients (implicitly) so that similar parts of the backfill are requested at similar times. Ultimately all data will have to be retrieved, but the hope is that we can minimize the costs to peers, and have Beam Syncing peers support each other more.
+
+The proposal is to backfill in this pattern:
+1. On the most recent block where `block.number % 100 == 0`, read the block hash and choose it as your backfill seed
+2. Treat the backfill seed as if it were a hashed address
+3. Use the seed to index into the trie, and start an upward scan of accounts and storage (collecting all storage for one account before moving to the next sibling account)
```

Comment:
Maybe worth adding a note here or below about wrapping around when we reach the `0xffffffff....` location.